compiler/src/iree/compiler/GlobalOptimization/SetEncoding.cpp

```diff
@@ -75,19 +78,46 @@ static Value pad(OpBuilder &builder, Location loc, Value source,
                                        lowPad, highPad, zero);
 }
 
-static Value setEncoding(OpBuilder &builder, Location loc, Value source,
-                         IREE::LinalgExt::EncodingAttr encodingAttr) {
+Value setEncoding(OpBuilder &builder, Location loc, Value source,
+                  IREE::LinalgExt::EncodingAttr encodingAttr) {
   auto sourceType = source.getType().cast<RankedTensorType>();
   auto resultType = RankedTensorType::get(
       sourceType.getShape(), sourceType.getElementType(), encodingAttr);
   return builder.create<IREE::LinalgExt::SetEncodingOp>(loc, resultType,
                                                         source);
 };
 
-static IREE::LinalgExt::EncodingAttr
+struct MatmulNarrowSizes {
+  std::optional<int64_t> M, N;
+};
+
+// Returns the minimum of static sizes of the M-dimension in the types of the
+// LHS and/or the Output operand of a matmul, whichever is static.
+MatmulNarrowSizes getMatmulNarrowSizes(ShapedType outType) {
+  int64_t M = outType.getDimSize(0);
+  int64_t N = outType.getDimSize(1);
+  MatmulNarrowSizes narrow;
+  // Threshold below which a M/N size is considered "narrow", making it
+  // eligible for a narrow tile size during materialization. This value should
+  // be at least as large as the actual M/N tile sizes that we choose on any
+  // target in CPUMaterializeEncodingPass. If it is smaller, we will miss
+  // opportunities to select optimized narrow tiles for narrow matmuls.
+  // If it is larger, everything will work fine, but the IR will be a bit more
+  // verbose as more narrow_matmul_{M,N} optional parameters will be specified.
+  const int64_t narrowThreshold = 16;
+  if (M != ShapedType::kDynamic && M < narrowThreshold) {
```

Comment:
nit: use `ShapedType::isDynamic(M)`. I recently learned that one of MLIR goals is to make ShapedType::kDynamic private, and force all users to go through safe APIs. Let's use it for new codes. Same for the below `N` case.

```suggestion
  if (!ShapedType::isDynamic(M) && M < narrowThreshold) {
```

(just in case, https://discourse.llvm.org/t/rfc-unify-kdynamicsize-and-kdynamicstrideoroffset/64534 has a little context. What I learned is more from review comments: https://reviews.llvm.org/D138282)

Reply:
>  I recently learned that one of MLIR goals is to make ShapedType::kDynamic private, and force all users to go through safe APIs.

MLIR fills me with hope: maybe it is not inevitable that we all grow into old, cynical, experienced C++ developers. Maybe we can keep believing that we can make the world a better place by hiding integer constants that are aliases for INT64_MIN.